### PR TITLE
feature: add --format option for pouch ps

### LINKED ISSUE #2827: [bug] cli compatibility between Pouch and Docker
### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1.
2.
3.

### Ⅵ. Environment:

- pouch version (use `pouch version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by aohang111**
### Ⅰ. Describe what this PR did
for the issue #2827. CLI compatibility between Pouch and Docker 

solve the pouch ps --format option.
if other command(such as pouch history/images/network/volume/stats/version) also need --format option.
can done similarly

### Ⅱ. Does this pull request fix one issue?

fixes #2827,

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

have add unit test and integration test

### Ⅳ. Describe how to verify it

root@ahsh-main:~/go/src/github.com/alibaba/pouch# pouch ps
Name   ID   Status   Created   Image   Runtime
58b7de   58b7de   Up 40 minutes   40 minutes ago   registry.hub.docker.com/library/ubuntu:latest   runc

root@ahsh-main:~/go/src/github.com/alibaba/pouch# pouch ps -a --format "table {{.ID}}\t{{.Status}}\t{{.Created}}"
ID   Status   Created
58b7de   Up 11 seconds   11 seconds ago

root@ahsh-main:~/go/src/github.com/alibaba/pouch# pouch ps -a --format "table {{.ID}}\t{{.Status}}\t{{.Created}}\t{{.Name}}"
ID   Status   Created   Name
58b7de   Up 40 minutes   40 minutes ago   58b7de

root@ahsh-main:~/go/src/github.com/alibaba/pouch# pouch ps -a --format "raw"
Name:58b7de
ID:58b7de
Status:Up 4 seconds
Created:4 seconds ago
Image:registry.hub.docker.com/library/ubuntu:latest
Runtime:runc

### Ⅴ. Special notes for reviews

with @houstar's help.

### CODE REVIEW

**Comment by houstar** [comment]
@aohang111  Could you use gofmt -s to format your source code as officially style? thanks

**Comment by allencloud** [comment]
Thank you greatly for your work. @aohang111 
While I am wondering if we could add integration test for the pull request, not only the unit test. 
Please feel free to tell us if you have some questions. Thanks again.

**Comment by houstar** [comment]
Could you squash previous four commits into one commits ?

**Comment by aohang111** [comment]
> Please feel free to tell us if you have some questions. Thanks again.
I am trying to add the integration test. by the way. if the scripts are in dir test. I will use the scripts in that dir as a guide.

**Comment by aohang111** [comment]
@allencloud ,I have added  the integration test. please assign someone to review this code ASAP

**Comment by houstar** [comment]
Well done. @aohang111  You've made this happen as your first contribution to pouch project

File: cli/cli.go [hunk 1]
```diff
@@ -159,6 +160,20 @@ func (c *Cli) Print(obj interface{}) {
 	display.Flush()
 }
 
+//FormatDisplay use to format output with options
```
Comment (ZYecho) [inline cli/cli.go]: add a blank space.

File: cli/formatter/container.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package formatter
+
+import (
+	"github.com/alibaba/pouch/apis/types"
+	"github.com/alibaba/pouch/pkg/utils"
+)
+
+//ContainerHeader is the map to show container head
```
Comment (ZYecho) [inline cli/formatter/container.go]: add a blank space. please modify this problem In this pr.

File: cli/formatter/formatter_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			output: "{{.ID}}\t{{.Status}}\t{{.Created}}\t{{.Name}}\t{{.Image}}\t{{.State}}\n",
+		},
+		{
+			input:  "table {{.ID}}\t{{.Status}}\t{{.Created}}\\t{{.Name}}\t{{.Image}}\t{{.State}}\n",
+			output: "{{.ID}}\t{{.Status}}\t{{.Created}}\t{{.Name}}\t{{.Image}}\t{{.State}}\n",
+		},
+		{
+			input: "raw",
+			output: `Name:{{.Name}}
+ID:{{.ID}}
```
Comment (ZYecho) [inline cli/formatter/formatter_test.go]: please format this part, looks a little ugly.

File: cli/formatter/formatter.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		//input by cmd of "\t" is "\\t"
+		replace := strings.NewReplacer(`\t`, "\t", `\n`, "\n")
+		format = replace.Replace(format)
+
+		if format[len(format)-1:] != "\n" {
+			format += "\n"
+		}
+	} else {
+		format = `Name:{{.Name}}
+ID:{{.ID}}
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: please format this part.

File: cli/formatter/formatter.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package formatter
+
+import (
+	"sort"
+	"strings"
+)
+
+//Format key to output
+const (
+	TableFormat = "table"
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: The difference between table and raw is that raw's output includes the container header, right?

File: cli/formatter/container.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+type ContainerHeader map[string]string
+
+//ContainerContext is the map to show container context detail
+type ContainerContext map[string]string
+
+//NewContainerHeader is to generate a ContainerHeader to be show
+func NewContainerHeader() ContainerHeader {
```
Comment (ZYecho) [inline cli/formatter/container.go]: seem we define a const map here is enough, no need to define a func?

File: cli/formatter/container.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		"ImageID": "ImageID",
+		"Labels":  "Labels",
+		//"Mounts":  "Mounts",
+		"State": "State",
+	}
+	return containerHeader
+}
+
+//NewContainerContext is to generate a ContainerContext to be show
+func NewContainerContext(c *types.Container, flagNoTrunc bool) ContainerContext {
```
Comment (ZYecho) [inline cli/formatter/container.go]: the container context's key is same with the container head?

File: cli/formatter/formatter.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+//LabelsToString is to transfor the labels from map to string
+func LabelsToString(labels map[string]string) string {
+	var labelstring string
+	sortedkeys := make([]string, 0)
+	for key := range labels {
+		sortedkeys = append(sortedkeys, key)
+	}
+	sort.Strings(sortedkeys)
+	for _, key := range sortedkeys {
+		keyValueString := key + "=" + labels[key] + " "
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: better to use `labelstring += fmt.Sprintf("%s = %s", key, value)`

File: cli/formatter/container.go (hunk 7 above)
Comment (ZYecho) [inline cli/formatter/container.go]: I think `ToContainerContext` is a better name to understand.

File: cli/formatter/formatter.go (hunk 5 above)
Comment (aohang111) [reply cli/formatter/formatter.go]: yes, if using table, the first line should be the container header. like this:
pouch ps -a --format "table {{.ID}}\t{{.Status}}\t{{.Created}}\t{{.Name}}"
ID   Status   Created   Name
f78484   Up 32 seconds   33 seconds ago   f78484
58b7de   Up 3 days   3 days ago   58b7de

if using raw, result like below:
pouch ps -a --format "raw"
Name:f78484
ID:f78484
Status:Up 13 seconds
Created:13 seconds ago
Image:registry.hub.docker.com/library/ubuntu:latest
Runtime:runc

Name:58b7de
ID:58b7de
Status:Up 3 days
Created:3 days ago
Image:registry.hub.docker.com/library/ubuntu:latest
Runtime:runc

File: cli/formatter/container.go (hunk 7 above)
Comment (aohang111) [reply cli/formatter/container.go]: yes, they have the same keys.

File: cli/formatter/formatter.go (hunk 8 above)
Comment (aohang111) [reply cli/formatter/formatter.go]: good suggestion.

File: cli/formatter/formatter.go (hunk 4 above)
Comment (aohang111) [reply cli/formatter/formatter.go]: done

File: cli/formatter/container.go (hunk 6 above)
Comment (aohang111) [reply cli/formatter/container.go]: yes , you are right. will change the func to const

File: cli/formatter/container.go (hunk 7 above)
Comment (aohang111) [reply cli/formatter/container.go]: my workflow of this issue is: 
1,we need the template package to show the format result.
2,we must have datastruct to show .
3,we have two kind of data(container head and container context)
4,we build the two kinds of data(using two func.)

so I thought we can use NewContainerContext/BuildContainerContext.

what's your concern about using  "ToContainerContext"

**Comment by aohang111** [comment]
@ZYecho ,modified and comments. pls review. thanks

File: cli/formatter/container.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	"Name":    "Name",
+	"ID":      "ID",
+	"Status":  "Status",
+	"Created": "Created",
+	"Image":   "Image",
+	"Runtime": "Runtime",
+	"Command": "Command",
+	"ImageID": "ImageID",
+	"Labels":  "Labels",
+	// "Mounts":  "Mounts",
```
Comment (ZYecho) [inline cli/formatter/container.go]: should we add a comment for this？ what's your concern？

File: cli/formatter/formatter.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+// IsTable to verify if table or raw
+func IsTable(format string) bool {
+	return strings.HasPrefix(format, TableFormat)
+}
+
+// PreFormat is to format the format option
+func PreFormat(format string) string {
+	if IsTable(format) {
+		format = format[len(TableFormat):]
+		//cut the space
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: add a blank space.

File: cli/formatter/formatter.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+		format = strings.Trim(format, " ")
+		//input by cmd of "\t" is "\\t"
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: add a blank space.

File: cli/formatter/formatter.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+		replace := strings.NewReplacer(`\t`, "\t", `\n`, "\n")
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: could you explain the logic here? why we need this? thanks.

File: cli/formatter/formatter.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+		if format[len(format)-1:] != "\n" {
+			format += "\n"
+		}
+	} else {
+		format = "Name:{{.Name}}\nID:{{.ID}}\nStatus:{{.Status}}\nCreated:{{.Created}}\nImage:{{.Image}}\nRuntime:{{.Runtime}}\n\n"
+	}
+	return format
+}
+
+// LabelsToString is to transfor the labels from map to string
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: transfor -> transform?

**Comment by ZYecho** [comment]
> @ZYecho ,modified and comments. pls review. thanks

Thanks for your contribution, just minor suggestions. Good Job!

File: cli/formatter/container.go (hunk 9 above)
Comment (aohang111) [reply cli/formatter/container.go]: no need, I can add the mount data.

File: cli/formatter/formatter.go (hunk 12 above)
Comment (aohang111) [reply cli/formatter/formatter.go]: format = strings.Trim(format, " ")
user input "table        {{.ID}}         "------ no trim for space:the output will have so many space
replace := strings.NewReplacer(`\t`, "\t", `\n`, "\n")
seems in the cobra command. the input of "\" will be consider as `\`  not "\".
so need to transfer it.

**Comment by aohang111** [comment]
@ZYecho commit again. pls review. thanks

File: test/cli_ps_test.go [hunk 14]
```diff
@@ -256,6 +256,48 @@ func (suite *PouchPsSuite) TestPsNoTrunc(c *check.C) {
...
+
+	resRaw := command.PouchRun("ps", "--format", "raw").Assert(c, icmd.Success)
+	resRawHead := findRawFormatPsHead(resRaw.Combined())
+	c.Assert(len(resRawHead), check.Equals, 7)
+	c.Assert(resRawHead[0], check.Equals, "Name")
+	c.Assert(resRawHead[1], check.Equals, "ID")
+	c.Assert(resRawHead[2], check.Equals, "Status")
+	c.Assert(resRawHead[3], check.Equals, "Created")
+	c.Assert(resRawHead[4], check.Equals, "Image")
+	c.Assert(resRawHead[5], check.Equals, "Runtime")
```
Comment (ZYecho) [inline test/cli_ps_test.go]: need to check mounts filed？

File: cli/formatter/formatter.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	sort.Strings(sortedkeys)
+	for _, key := range sortedkeys {
+		labelstring += fmt.Sprintf("%s = %s;", key, labels[key])
+	}
+	return labelstring
+}
+
+// MountPointToString is to transform the MountPoint from array to string
+func MountPointToString(mount []types.MountPoint) string {
```
Comment (ZYecho) [inline cli/formatter/formatter.go]: could you add a unit test for this func?

File: test/cli_ps_test.go (hunk 14 above)
Comment (aohang111) [reply test/cli_ps_test.go]: yes. will add later

File: cli/formatter/formatter.go (hunk 15 above)
Comment (aohang111) [reply cli/formatter/formatter.go]: done

**Comment by aohang111** [comment]
@ZYecho .done.

**Comment by ZYecho** [comment]
LGTM

**Comment by aohang111** [comment]
@ZYecho so, what the next step.what should I do about the PR.

**Comment by ZYecho** [comment]
@aohang111 waiting another maintainer to give you an LGTM. @allencloud  could you help to take a look?

**Comment by aohang111** [comment]
@ZYecho ,just waiting...

File: cli/cli.go [hunk 16]
```diff
@@ -159,6 +160,18 @@ func (c *Cli) Print(obj interface{}) {
 	display.Flush()
 }
 
+// FormatDisplay use to format output with options
+func (c *Cli) FormatDisplay(format string, tmpl *template.Template, container interface{}, w *tabwriter.Writer) error {
+	tmpl, err := tmpl.Parse(format)
+	if err != nil {
+		return err
+	}
+	if err := tmpl.Execute(w, container); err != nil {
```
Comment (allencloud) [inline cli/cli.go]: could you just use `return tmpl.Execute(w, container)` to simplify code?

File: cli/formatter/container.go [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+		id = c.ID
+	}
+	created, _ := utils.FormatTimeInterval(c.Created)
+	labels := LabelsToString(c.Labels)
+	mount := MountPointToString(c.Mounts)
+	containerContext := ContainerContext{
+		"Name":    c.Names[0],
+		"ID":      id,
+		"Status":  c.Status,
+		"Created": created + " ago",
```
Comment (allencloud) [inline cli/formatter/container.go]: please use `fmt.Sprintf("%s ago", created)` instead of `created + " ago"`.

**Comment by allencloud** [comment]
I have tested towards docker engine:
 ```
~ # docker ps -a --format "table {{.Created}}"
Template parsing error: template: :1:2: executing "" at <.Created>: can't evaluate field Created in type *formatter.containerContext
```
Failed.

Is there any incompatibility or some other issue?   @aohang111

**Comment by aohang111** [comment]
> I have tested towards docker engine:
> 
> ```
> ~ # docker ps -a --format "table {{.Created}}"
> Template parsing error: template: :1:2: executing "" at <.Created>: can't evaluate field Created in type *formatter.containerContext
> ```
> 
> Failed.
> 
> Is there any incompatibility or some other issue? @aohang111

from the code, seems need to using this
ps -a --format "table {{.CreatedAt}}"
CREATED AT
2019-05-17 03:01:07 +0000 UTC
2019-05-17 03:00:49 +0000 UTC

File: cli/formatter/container.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+// ContainerContext is the map to show container context detail
+type ContainerContext map[string]string
+
+// NewContainerContext is to generate a ContainerContext to be show
+func NewContainerContext(c *types.Container, flagNoTrunc bool) ContainerContext {
+	id := c.ID[:6]
+	if flagNoTrunc {
+		id = c.ID
+	}
+	created, _ := utils.FormatTimeInterval(c.Created)
```
Comment (fuweid) [inline cli/formatter/container.go]: could we check the error here? I think the original code follows the error check here.

File: cli/ps.go [hunk 19]
```diff
@@ -88,24 +94,26 @@ func (p *PsCommand) runPs(args []string) error {
 		}
 		return nil
 	}
-
-	display := p.cli.NewTableDisplay()
-	display.AddRow([]string{"Name", "ID", "Status", "Created", "Image", "Runtime"})
-
+	// add to format the output with go template
+	format := p.flagFormat
```
Comment (fuweid) [inline cli/ps.go]: I think we can refactor code to include the quiet mode. I am fine to defer to discuss with this.

Comment (aohang111) [reply cli/ps.go]: if it is quiet mode, we have no need to care about the format option.

File: cli/formatter/container.go (hunk 18 above)
Comment (aohang111) [reply cli/formatter/container.go]: done

**Comment by aohang111** [comment]
@allencloud ,have add all format key which docker have.
@fuweid feel free to give your suggestion about the quiet mode.

**Comment by houstar** [comment]
Any progress here?

**Comment by ZYecho** [comment]
@fuweid PTAL, thanks.

**Comment by aohang111** [comment]
@fuweid ,if the user specify the quite mode. we just need to output the shorted container id . no need to care about if the format is specified. so I thought it is split with the logic of format display. what's your opinion .

**Comment by aohang111** [comment]
@allencloud @ZYecho @fuweid.

**Comment by aohang111** [comment]
@allencloud @ZYecho @fuweid @houstar  ,what shall I do with this PR.

**Comment by cxz66666** [comment]
> @fuweid ,if the user specify the quite mode. we just need to output the shorted container id . no need to care about if the format is specified. so I thought it is split with the logic of format display. what's your opinion .

I also think it's needless to care about the output format in quite mode, please resolve some conflicts ASAP and latter it will be merged into master!

File: cli/ps.go [hunk 20]
```diff
@@ -88,24 +94,29 @@ func (p *PsCommand) runPs(args []string) error {
...
-		if p.flagNoTrunc {
-			id = c.ID
-		}
-
-		display.AddRow([]string{c.Names[0], id, c.Status, created + " ago", c.Image, c.HostConfig.Runtime})
+		tmplD := template.New("ps_detail")
+		p.cli.FormatDisplay(format, tmplD, containerContext, w)
 	}
-	display.Flush()
+	w.Flush()
```
Comment (cxz66666) [inline cli/ps.go:119]: please use ```_=w.Flush()``` instead of it!

**Comment by cxz66666** [comment]
Thank you for your contribution! because you don't response for a long time, I will close this PR and reopen https://github.com/alibaba/pouch/pull/3037. ❤️

**Pull request closed without merging**